backend/fileSystem.js

```diff
@@ -145,6 +147,84 @@ class FileSystemBackend {
         const match = await this.getPackageByName(packageName);
         return match ? match.versions[version] : null;
     }
+    async searchPackages(properties) {
+        // Retrieve listing of all packages
+        await this.syncMetadata();
+        const packages = Object.keys(this._metadata).map(key => this._metadata[key]);
+
+        // Begin with all packages
+        let result = packages;
+
+        // Filter by search query
+        if (typeof properties.query === 'string' && properties.query.length > 0) {
+            // If only searching for names, it's possible to more efficiently search the whole package array
+            let searchProperties = packages;
+
+            // If doing an advanced search with additional properties, make a new array of package properties
+            if (properties.property.length !== 1 || properties.property[0] !== 'name') {
```

Comment:
What if properties.property.length === 0? Looks like this may throw an undefined error. I think it would be better to either check for `properties.property.length` or `properties.property.length > 0` explicitly.

Also, can you explain what would happen if the OR statement were to pass as "true," but 'name' occurred in the properties.property array at some other index?